File: analysis/discover_apify_actors.py
#!/usr/bin/env python3
"""
Apify Actor Discovery Script
Discovers all Apify actors, analyzes their schemas, and generates comprehensive inventory report
"""

import os
import json
import requests
from datetime import datetime
from pathlib import Path
from typing import List, Dict, Any
from dotenv import load_dotenv

# Load environment variables
env_path = Path(__file__).parent.parent / '.env'
load_dotenv(env_path)

# Configuration
APIFY_API_KEY = os.getenv('APIFY_API_KEY', '')
COMPOSIO_MCP_URL = os.getenv('COMPOSIO_MCP_URL', 'http://localhost:3001')
UNIQUE_ID = '04.01.99.11.05000.001'
PROCESS_ID = 'Apify Actor Discovery and Schema Analysis'

# Known actors to check
KNOWN_ACTORS = [
    'apify/email-phone-scraper',
    'apify/linkedin-profile-scraper',
    'apify/website-content-crawler',
    'apify/linkedin-company-scraper',
    'apify/google-search-scraper',
    'apify/contact-finder',
    'apify/people-data-scraper',
    'apify/web-scraper',
    'compass/apollo-scraper',
    'compass/linkedin-people-scraper',
]

# Executive enrichment keywords
EXECUTIVE_KEYWORDS = [
    'ceo', 'cfo', 'hr', 'executive', 'officer', 'director', 'manager',
    'leadership', 'title', 'role', 'position', 'linkedin', 'contact',
    'email', 'name', 'company', 'organization', 'people', 'profile'
]

def fetch_actor_details(actor_id: str) -> Dict[str, Any]:
    """Fetch detailed information about an Apify actor"""
    url = f'https://api.apify.com/v2/acts/{actor_id}'
    params = {'token': APIFY_API_KEY}

    print(f'  [FETCH] Fetching: {actor_id}')

    try:
        response = requests.get(url, params=params, timeout=10)

        if response.status_code == 200:
            data = response.json()
            print(f'     [OK] Success')
            return data.get('data', {})
        else:
            print(f'     [WARN] HTTP {response.status_code}')
            return None
    except Exception as e:
        print(f'     [ERROR] Error: {str(e)}')
        return None

def analyze_for_executive_enrichment(actor: Dict[str, Any]) -> Dict[str, Any]:
    """Analyze an actor's relevance for executive enrichment"""

    title = (actor.get('title', '') or '').lower()
    description = (actor.get('description', '') or '').lower()
    input_schema = actor.get('inputSchema', {})

    score = 0
    relevant_fields = []

    # Analyze input schema properties
    properties = input_schema.get('properties', {})
    for field, field_data in properties.items():
        field_lower = field.lower()
        field_desc = (field_data.get('description', '') or '').lower()

        for keyword in EXECUTIVE_KEYWORDS:
            if keyword in field_lower or keyword in field_desc:
                score += 10
                relevant_fields.append({
                    'field': field,
                    'type': field_data.get('type', 'unknown'),
                    'description': field_data.get('description', ''),
                    'keyword': keyword
                })
                break  # Count each field only once

    # Analyze title and description
    for keyword in EXECUTIVE_KEYWORDS:
        if keyword in title:
            score += 5
        if keyword in description:
            score += 3

    # Categorize actor
    is_people = any(kw in title or kw in description for kw in ['people', 'contact', 'profile', 'person'])
    is_company = any(kw in title or kw in description for kw in ['company', 'organization', 'business'])
    is_linkedin = 'linkedin' in title or 'linkedin' in description
    is_email = 'email' in title or 'email' in description

    return {
        'relevanceScore': score,
        'relevantFields': relevant_fields,
        'isPeopleEnrichment': is_people,
        'isCompanyEnrichment': is_company,
        'isLinkedInSpecific': is_linkedin,
        'isEmailFocused': is_email,
        'categories': {
            'people': is_people,
            'company': is_company,
            'linkedin': is_linkedin,
            'email': is_email
        }
    }

def extract_input_fields(input_schema: Dict[str, Any]) -> List[Dict[str, Any]]:
    """Extract and summarize input fields from schema"""
    properties = input_schema.get('properties', {})
    required = input_schema.get('required', [])

    fields = []
    for field_name, field_data in properties.items():
        fields.append({
            'name': field_name,
            'type': field_data.get('type', 'unknown'),
            'description': field_data.get('description', ''),
            'required': field_name in required,
            'default': field_data.get('default'),
            'enum': field_data.get('enum')
        })

    return fields

def extract_output_fields(output_schema: Dict[str, Any]) -> List[Dict[str, Any]]:
    """Extract and summarize output fields from schema"""
    properties = output_schema.get('properties', {})

    fields = []
    for field_name, field_data in properties.items():
        fields.append({
            'name': field_name,
            'type': field_data.get('type', 'unknown'),
            'description': field_data.get('description', '')
        })

    return fields

def discover_actors() -> Dict[str, Any]:
    """Main discovery function"""

    print('\n>> Starting Apify Actor Discovery')
    print('=' * 70)
    print(f'[KEY] API Key: {APIFY_API_KEY[:10]}...' if APIFY_API_KEY else '[ERROR] No API Key')
    print(f'[PROC] Process ID: {PROCESS_ID}')
    print(f'[ID] Unique ID: {UNIQUE_ID}')
    print('=' * 70)
    print()

    if not APIFY_API_KEY:
        print('[ERROR] APIFY_API_KEY not set in environment')
        print('   Please set it in .env file')
        return None

    results = {
        'timestamp': datetime.utcnow().isoformat() + 'Z',
        'unique_id': UNIQUE_ID,
        'process_id': PROCESS_ID,
        'discoveredActors': [],
        'currentlyUsedActors': [
            {
                'actorId': 'apify/email-phone-scraper',
                'usage': 'Primary contact scraper in apifyRunner.js',
                'location': 'agents/specialists/apifyRunner.js:18'
            },
            {
                'actorId': 'apify~linkedin-profile-scraper',
                'usage': 'LinkedIn profile scraping',
                'location': 'packages/mcp-clients/src/clients/apify-mcp-client.ts:57'
            },
            {
                'actorId': 'apify~website-content-crawler',
                'usage': 'Website contact extraction',
                'location': 'packages/mcp-clients/src/clients/apify-mcp-client.ts:118'
            }
        ],
        'mockActors': [
            'linkedin-company-scraper',
            'website-content-crawler',
            'company-data-scraper',
            'business-permit-scraper',
            'financial-data-scraper'
        ],
        'errors': []
    }

    print('[LIST] Discovering Apify actors...\n')

    for actor_id in KNOWN_ACTORS:
        try:
            actor_data = fetch_actor_details(actor_id)

            if actor_data:
                # Analyze for executive enrichment
                analysis = analyze_for_executive_enrichment(actor_data)

                # Extract input/output schemas
                input_fields = extract_input_fields(actor_data.get('inputSchema', {}))
                output_fields = extract_output_fields(actor_data.get('outputSchema', {}))

                # Build actor record
                actor_record = {
                    'actorId': actor_id,
                    'title': actor_data.get('title', ''),
                    'description': actor_data.get('description', '')[:500],
                    'username': actor_data.get('username', actor_id.split('/')[0]),
                    'isPublic': actor_data.get('isPublic', False),
                    'stats': {
                        'totalRuns': actor_data.get('stats', {}).get('totalRuns', 0),
                        'totalUsers': actor_data.get('stats', {}).get('totalUsers', 0)
                    },
                    'inputFields': input_fields,
                    'outputFields': output_fields,
                    'analysis': analysis,
                    'versions': actor_data.get('versions', [])
                }

                results['discoveredActors'].append(actor_record)

                print(f'     [SCORE] Relevance Score: {analysis["relevanceScore"]}')
                print(f'     [PEOPLE] People: {analysis["isPeopleEnrichment"]}')
                print(f'     [COMPANY] Company: {analysis["isCompanyEnrichment"]}')
                print(f'     [LINKEDIN] LinkedIn: {analysis["isLinkedInSpecific"]}')
                print(f'     [EMAIL] Email: {analysis["isEmailFocused"]}')
                print(f'     [INPUT] Input Fields: {len(input_fields)}')
                print(f'     [OUTPUT] Output Fields: {len(output_fields)}')
                print()

        except Exception as e:
            results['errors'].append({
                'actorId': actor_id,
                'error': str(e)
            })
            print()

    # Rank actors by relevance
    ranked = sorted(
        [a for a in results['discoveredActors'] if a['analysis']['relevanceScore'] > 0],
        key=lambda x: x['analysis']['relevanceScore'],
        reverse=True
    )

    results['topExecutiveEnrichmentActors'] = [
        {
            'rank': i + 1,
            'actorId': actor['actorId'],
            'title': actor['title'],
            'relevanceScore': actor['analysis']['relevanceScore'],
            'categories': actor['analysis']['categories'],
            'keyInputFields': [f['name'] for f in actor['input Fields'][:5]],
            'keyOutputFields': [f['name'] for f in actor['outputFields'][:5]]
        }
        for i, actor in enumerate(ranked[:10])
    ]

    return results

def save_results(results: Dict[str, Any]):
    """Save discovery results to file"""
    output_dir = Path(__file__).parent
    output_file = output_dir / 'apify_discovery_raw.json'

    with open(output_file, 'w', encoding='utf-8') as f:
        json.dump(results, f, indent=2, ensure_ascii=False)

    print(f'[SAVE] Saved raw results to: {output_file}')
    return output_file

def print_summary(results: Dict[str, Any]):
    """Print discovery summary"""
    print('\n[SUMMARY] DISCOVERY SUMMARY')
    print('=' * 70)
    print(f"Total Actors Discovered: {len(results['discoveredActors'])}")
    print(f"Currently Used in Codebase: {len(results['currentlyUsedActors'])}")
    print(f"Errors Encountered: {len(results['errors'])}")

    print('\n[TOP] Top 3 Actors for Executive Enrichment:\n')

    for actor in results['topExecutiveEnrichmentActors'][:3]:
        print(f"{actor['rank']}. {actor['actorId']}")
        print(f"   [SCORE] Score: {actor['relevanceScore']}")
        print(f"   [TITLE] Title: {actor['title']}")
        print(f"   [PEOPLE] People: {actor['categories']['people']}")
        print(f"   [COMPANY] Company: {actor['categories']['company']}")
        print(f"   [LINKEDIN] LinkedIn: {actor['categories']['linkedin']}")
        print(f"   [EMAIL] Email: {actor['categories']['email']}")
        if actor.get('keyInputFields'):
            print(f"   [INPUTS] Key Inputs: {', '.join(actor['keyInputFields'])}")
        print()

    print('=' * 70)

if __name__ == '__main__':
    results = discover_actors()

    if results:
        output_file = save_results(results)
        print_summary(results)
        print(f'\n[OK] Discovery completed successfully!')
        print(f'[FILE] Results saved to: {output_file}')
    else:
        print('\n[ERROR] Discovery failed')
        exit(1)

File: ctb/docs/analysis/discover_apify_actors.js
/*
─────────────────────────────────────────────
📁 CTB Classification Metadata
─────────────────────────────────────────────
CTB Branch: docs/analysis
Barton ID: 06.01.01
Unique ID: CTB-8FCD3F43
Blueprint Hash:
Last Updated: 2025-10-23
Enforcement: ORBT
─────────────────────────────────────────────
*/

/**
 * Apify Actor Discovery Script
 * Connects to Composio MCP to discover all Apify actors and their schemas
 */

import fetch from 'node-fetch';
import fs from 'fs/promises';
import path from 'path';

const COMPOSIO_MCP_URL = process.env.COMPOSIO_MCP_URL || 'http://localhost:3001';
const COMPOSIO_API_KEY = process.env.COMPOSIO_API_KEY || 'ak_t-F0AbvfZHUZSUrqAGNn';
const APIFY_API_KEY = process.env.APIFY_API_KEY;

// Barton Doctrine metadata
const UNIQUE_ID = '04.01.99.11.05000.001';
const PROCESS_ID = 'Apify Actor Discovery and Schema Analysis';

/**
 * Call Composio MCP tool
 */
async function callComposioTool(tool, params, metadata = {}) {
  const payload = {
    tool,
    data: params,
    unique_id: UNIQUE_ID,
    process_id: PROCESS_ID,
    orbt_layer: 5,
    blueprint_version: '1.0',
    ...metadata
  };

  console.log(`📡 Calling Composio tool: ${tool}`);

  const response = await fetch(`${COMPOSIO_MCP_URL}/tool`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      'X-Composio-Api-Key': COMPOSIO_API_KEY
    },
    body: JSON.stringify(payload)
  });

  if (!response.ok) {
    const errorText = await response.text();
    throw new Error(`Composio MCP call failed: ${response.statusText} - ${errorText}`);
  }

  return await response.json();
}

/**
 * Fetch actor details directly from Apify API
 */
async function fetchActorDetails(actorId) {
  if (!APIFY_API_KEY) {
    throw new Error('APIFY_API_KEY not configured');
  }

  const url = `https://api.apify.com/v2/acts/${actorId}?token=${APIFY_API_KEY}`;

  console.log(`  📥 Fetching details for: ${actorId}`);

  const response = await fetch(url);

  if (!response.ok) {
    console.warn(`  ⚠️  Failed to fetch ${actorId}: ${response.statusText}`);
    return null;
  }

  return await response.json();
}

/**
 * Analyze actor input schema for executive enrichment relevance
 */
function analyzeActorForExecutives(actor) {
  const schema = actor.data || actor;
  const inputSchema = schema.input_schema || schema.inputSchema || {};
  const defaultInput = schema.default_input || schema.defaultInput || {};

  const executiveKeywords = [
    'ceo', 'cfo', 'hr', 'executive', 'officer', 'director', 'manager',
    'leadership', 'title', 'role', 'position', 'linkedin', 'contact',
    'email', 'name', 'company', 'organization'
  ];

  let score = 0;
  let relevantFields = [];

  // Analyze input schema properties
  const properties = inputSchema.properties || {};
  Object.keys(properties).forEach(field => {
    const fieldLower = field.toLowerCase();
    const fieldData = properties[field];
    const description = (fieldData.description || '').toLowerCase();

    executiveKeywords.forEach(keyword => {
      if (fieldLower.includes(keyword) || description.includes(keyword)) {
        score += 10;
        relevantFields.push({
          field,
          type: fieldData.type,
          description: fieldData.description || '',
          keyword
        });
      }
    });
  });

  // Check title and description
  const title = (schema.title || '').toLowerCase();
  const description = (schema.description || '').toLowerCase();

  executiveKeywords.forEach(keyword => {
    if (title.includes(keyword)) score += 5;
    if (description.includes(keyword)) score += 3;
  });

  return {
    relevanceScore: score,
    relevantFields,
    isPeopleEnrichment: title.includes('people') || title.includes('contact') || title.includes('profile'),
    isCompanyEnrichment: title.includes('company') || title.includes('organization') || title.includes('business'),
    isLinkedInSpecific: title.includes('linkedin') || description.includes('linkedin')
  };
}

/**
 * Main discovery function
 */
async function discoverApifyActors() {
  console.log('\n🚀 Starting Apify Actor Discovery');
  console.log('═'.repeat(70));
  console.log(`📍 Composio MCP: ${COMPOSIO_MCP_URL}`);
  console.log(`🔑 Process ID: ${PROCESS_ID}`);
  console.log(`🆔 Unique ID: ${UNIQUE_ID}`);
  console.log('═'.repeat(70));

  const results = {
    timestamp: new Date().toISOString(),
    unique_id: UNIQUE_ID,
    process_id: PROCESS_ID,
    discoveredActors: [],
    currentlyUsedActors: [
      {
        actorId: 'apify/email-phone-scraper',
        usage: 'Primary contact scraper in apifyRunner.js',
        location: 'agents/specialists/apifyRunner.js:18'
      },
      {
        actorId: 'apify~linkedin-profile-scraper',
        usage: 'LinkedIn profile scraping',
        location: 'packages/mcp-clients/src/clients/apify-mcp-client.ts:57'
      },
      {
        actorId: 'apify~website-content-crawler',
        usage: 'Website contact extraction',
        location: 'packages/mcp-clients/src/clients/apify-mcp-client.ts:118'
      }
    ],
    mockActors: [
      'linkedin-company-scraper',
      'website-content-crawler',
      'company-data-scraper',
      'business-permit-scraper',
      'financial-data-scraper'
    ],
    errors: []
  };

  try {
    // Step 1: Try to list actors via Composio MCP
    console.log('\n📋 Step 1: Listing Apify actors via Composio MCP...\n');

    try {
      const actorListResult = await callComposioTool('apify.list_actors', {
        my: true,
        limit: 100
      });

      if (actorListResult.success && actorListResult.data) {
        console.log(`✅ Found ${actorListResult.data.items?.length || 0} actors via MCP`);
        if (actorListResult.data.items) {
          results.discoveredActors.push(...actorListResult.data.items);
        }
      }
    } catch (mcpError) {
      console.warn('⚠️  MCP list_actors not available, falling back to direct API');
      results.errors.push({
        step: 'list_actors_mcp',
        error: mcpError.message
      });
    }

    // Step 2: Fetch details for known actors
    console.log('\n📋 Step 2: Fetching details for known actors...\n');

    const knownActorIds = [
      'apify/email-phone-scraper',
      'apify/linkedin-profile-scraper',
      'apify/website-content-crawler',
      'apify/linkedin-company-scraper',
      'apify/google-search-scraper',
      'apify/contact-finder',
      'apify/people-data-scraper',
      'compass/apollo-scraper',
      'compass/linkedin-people-scraper'
    ];

    for (const actorId of knownActorIds) {
      try {
        const actorDetails = await fetchActorDetails(actorId);

        if (actorDetails) {
          // Analyze for executive enrichment
          const analysis = analyzeActorForExecutives(actorDetails);

          results.discoveredActors.push({
            actorId,
            title: actorDetails.data?.title || actorId,
            description: actorDetails.data?.description || '',
            username: actorDetails.data?.username || actorId.split('/')[0],
            isPublic: actorDetails.data?.isPublic !== false,
            stats: actorDetails.data?.stats || {},
            inputSchema: actorDetails.data?.inputSchema || {},
            outputSchema: actorDetails.data?.outputSchema || {},
            readme: actorDetails.data?.readme?.substr(0, 500) || '',
            analysis
          });

          console.log(`  ✅ ${actorId}`);
          console.log(`     📊 Relevance Score: ${analysis.relevanceScore}`);
          console.log(`     👥 People Enrichment: ${analysis.isPeopleEnrichment ? 'Yes' : 'No'}`);
          console.log(`     🏢 Company Enrichment: ${analysis.isCompanyEnrichment ? 'Yes' : 'No'}`);
          if (analysis.relevantFields.length > 0) {
            console.log(`     🎯 Key Fields: ${analysis.relevantFields.slice(0, 3).map(f => f.field).join(', ')}`);
          }
          console.log('');
        }
      } catch (error) {
        results.errors.push({
          actorId,
          step: 'fetch_details',
          error: error.message
        });
      }
    }

    // Step 3: Rank actors for executive enrichment
    console.log('\n🏆 Step 3: Ranking actors for executive enrichment...\n');

    const rankedActors = results.discoveredActors
      .filter(actor => actor.analysis && actor.analysis.relevanceScore > 0)
      .sort((a, b) => b.analysis.relevanceScore - a.analysis.relevanceScore)
      .slice(0, 10);

    results.topExecutiveEnrichmentActors = rankedActors.map((actor, index) => ({
      rank: index + 1,
      actorId: actor.actorId,
      title: actor.title,
      relevanceScore: actor.analysis.relevanceScore,
      isPeopleEnrichment: actor.analysis.isPeopleEnrichment,
      isLinkedInSpecific: actor.analysis.isLinkedInSpecific,
      keyFields: actor.analysis.relevantFields.slice(0, 5)
    }));

    // Step 4: Save results
    console.log('\n💾 Step 4: Saving discovery results...\n');

    const outputPath = path.join(process.cwd(), 'analysis', 'apify_discovery_raw.json');
    await fs.writeFile(outputPath, JSON.stringify(results, null, 2));
    console.log(`✅ Saved raw results to: ${outputPath}`);

    // Display summary
    console.log('\n📊 DISCOVERY SUMMARY');
    console.log('═'.repeat(70));
    console.log(`Total Actors Discovered: ${results.discoveredActors.length}`);
    console.log(`Currently Used in Codebase: ${results.currentlyUsedActors.length}`);
    console.log(`Errors Encountered: ${results.errors.length}`);
    console.log(`\nTop 3 Actors for Executive Enrichment:`);

    results.topExecutiveEnrichmentActors.slice(0, 3).forEach((actor, i) => {
      console.log(`\n${i + 1}. ${actor.actorId}`);
      console.log(`   📊 Score: ${actor.relevanceScore}`);
      console.log(`   👥 People-focused: ${actor.isPeopleEnrichment ? 'Yes' : 'No'}`);
      console.log(`   🔗 LinkedIn: ${actor.isLinkedInSpecific ? 'Yes' : 'No'}`);
      console.log(`   🎯 Key Fields: ${actor.keyFields.map(f => f.field).join(', ')}`);
    });

    console.log('\n═'.repeat(70));

    return results;

  } catch (error) {
    console.error('\n❌ Discovery failed:', error.message);
    console.error(error.stack);
    throw error;
  }
}

// Run if executed directly
if (import.meta.url === `file://${process.argv[1]}`) {
  discoverApifyActors()
    .then(() => {
      console.log('\n✅ Discovery completed successfully!');
      process.exit(0);
    })
    .catch(error => {
      console.error('\n💥 Discovery failed:', error.message);
      process.exit(1);
    });
}

export default discoverApifyActors;
